core/src/main/scala/org/apache/spark/deploy/history/HybridStore.scala

```diff
@@ -0,0 +1,185 @@
+  private var backgroundThread: Thread = null
+
+  // A hash map that stores all classes that had been writen to inMemoryStore
+  private val klassMap = new ConcurrentHashMap[Class[_], Boolean]
+
+  override def getMetadata[T](klass: Class[T]): T = {
+    getStore().getMetadata(klass)
+  }
+
+  override def setMetadata(value: Object): Unit = {
+    getStore().setMetadata(value)
+  }
+
+  override def read[T](klass: Class[T], naturalKey: Object): T = {
+    getStore().read(klass, naturalKey)
+  }
+
+  override def write(value: Object): Unit = {
+    getStore().write(value)
+
+    if (backgroundThread == null) {
+      // New classes won't be dumped once the background thread is started
+      klassMap.putIfAbsent(value.getClass(), true)
+    }
+  }
+
+  override def delete(klass: Class[_], naturalKey: Object): Unit = {
+    if (backgroundThread != null) {
+      throw new IllegalStateException("delete() shouldn't be called after " +
+        "the hybrid store begins switching to levelDB")
+    }
+
+    getStore().delete(klass, naturalKey)
+  }
+
+  override def view[T](klass: Class[T]): KVStoreView[T] = {
+    getStore().view(klass)
+  }
+
+  override def count(klass: Class[_]): Long = {
+    getStore().count(klass)
+  }
+
+  override def count(klass: Class[_], index: String, indexedValue: Object): Long = {
+    getStore().count(klass, index, indexedValue)
+  }
+
+  override def close(): Unit = {
+    try {
+      closed.set(true)
+      if (backgroundThread != null && backgroundThread.isAlive()) {
+        // The background thread is still running, wait for it to finish
+        backgroundThread.join()
+      }
+    } finally {
+      inMemoryStore.close()
+      if (levelDB != null) {
+        levelDB.close()
+      }
+    }
+  }
+
+  override def removeAllByIndexValues[T](
+      klass: Class[T],
+      index: String,
+      indexValues: Collection[_]): Boolean = {
+    if (backgroundThread != null) {
+      throw new IllegalStateException("removeAllByIndexValues() shouldn't be " +
+        "called after the hybrid store begins switching to levelDB")
+    }
+
+    getStore().removeAllByIndexValues(klass, index, indexValues)
+  }
+
+  def setLevelDB(levelDB: LevelDB): Unit = {
+    this.levelDB = levelDB
+  }
+
+  /**
+   * This method is called when the writing is done for inMemoryStore. A
+   * background thread will be created and be started to dump data in inMemoryStore
+   * to levelDB. Once the dumping is completed, the underlying kvstore will be
+   * switched to levelDB.
+   */
+  def switchToLevelDB(
+      listener: HybridStore.SwitchToLevelDBListener,
+      appId: String,
+      attemptId: Option[String]): Unit = {
+    if (closed.get) {
+      return
+    }
+
+    backgroundThread = new Thread(() => {
+      try {
+        for (klass <- klassMap.keys().asScala) {
+          val it = inMemoryStore.view(klass).closeableIterator()
+          while (it.hasNext()) {
+            levelDB.write(it.next())
+          }
```

Comment:
Drive by comment - given I added something similar to an in-house patch.
Add a `write(Iterator<E> values)` to kv store - this should make this switch order(s) faster.

Reply:
Thanks for the information, I will try adding a `write(Iterator<E> values)` in this pr.

Reply:
If this helps, this is what I had written up:
```
  @Override
  public void write(List<?> values) throws Exception {
    Preconditions.checkArgument(values != null && !values.isEmpty(),
      "Non-empty values required.");

    // Group by class, in case there are values from different classes in the iterator
    // Typical usecase is for this to be a single class.
    for (Map.Entry<? extends Class<?>, ? extends List<?>> entry :
            values.stream().collect(Collectors.groupingBy(Object::getClass)).entrySet()) {

      final Iterator<?> valueIter = entry.getValue().iterator();
      final Iterator<byte[]> serializedValueIter;

      {
        // deserialize outside synchronized block
        List<byte[]> list = new ArrayList<>(entry.getValue().size());
        for (Object value : entry.getValue()) {
          list.add(serializer.serialize(value));
        }
        serializedValueIter = list.iterator();
      }

      final Class<?> valueClass = entry.getKey();
      final LevelDBTypeInfo ti = getTypeInfo(valueClass);

      // Batching updates per type
      synchronized (ti) {
        final LevelDBTypeInfo.Index naturalIndex = ti.naturalIndex();
        final Collection<LevelDBTypeInfo.Index> indices = ti.indices();

        try (WriteBatch batch = db().createWriteBatch()) {
          while (valueIter.hasNext()) {
            final Object value = valueIter.next();

            assert serializedValueIter.hasNext();
            final byte[] serializedObject = serializedValueIter.next();

            Object existing;
            try {
              existing = get(naturalIndex.entityKey(null, value), valueClass);
            } catch (NoSuchElementException e) {
              existing = null;
            }

            PrefixCache cache = new PrefixCache(value);
            byte[] naturalKey = naturalIndex.toKey(naturalIndex.getValue(value));
            for (LevelDBTypeInfo.Index idx : indices) {
              byte[] prefix = cache.getPrefix(idx);
              idx.add(batch, value, existing, serializedObject, naturalKey, prefix);
            }
          }
          assert !serializedValueIter.hasNext();
          db().write(batch);
        }
      }
    }
  }

```

Reply:
Thanks, this is helpful!

Reply:
Hi @mridulm, I updated your code and used it on in-memory store - leveldb switching, but only saw little switching time improvement. I am not sure if somewhere wrong.
| log size, jobs and tasks per job            | 2 jobs, 400 tasks per job | 10 jobs, 400 tasks per job | 50 jobs, 400 tasks per job | 100 jobs, 400 tasks per job | 200 jobs, 400 tasks per job | 500 jobs, 400 tasks per job | 1000 jobs, 400 tasks per job | 5 jobs, 100000 tasks per job |
| ------------------------------------------- | ------------------------- | -------------------------- | -------------------------- | --------------------------- | --------------------------- | --------------------------- | ---------------------------- | ---------------------------- |
| original switching time                     | 1s                        | 2s                         | 4s                         | 8s                          | 16s                         | 37s                         | 65s                          | 90s                          |
| switching time with write(Iterator<T> iter) | 1s                        | 1s                         | 4s                         | 7s                          | 13s                         | 34s                         | 58s                          | 84s                          |

The code:
```
for (klass <- klassMap.keys().asScala) {
          val it = inMemoryStore.view(klass).closeableIterator()
          levelDB.write(it)
        }
```

```
    public <T> void write(Iterator<T> iter) throws Exception {
    Preconditions.checkArgument(iter != null, "Non-empty values required.");

    List<T> values = new ArrayList<>();
    iter.forEachRemaining(values::add);

    // Group by class, in case there are values from different classes in the iterator
    // Typical usecase is for this to be a single class.
    for (Map.Entry<? extends Class<?>, ? extends List<?>> entry :
            values.stream().collect(Collectors.groupingBy(Object::getClass)).entrySet()) {

      final Iterator<?> valueIter = entry.getValue().iterator();
      final Iterator<byte[]> serializedValueIter;

      {
        // deserialize outside synchronized block
        List<byte[]> list = new ArrayList<>(entry.getValue().size());
        for (Object value : entry.getValue()) {
          list.add(serializer.serialize(value));
        }
        serializedValueIter = list.iterator();
      }

      final Class<?> valueClass = entry.getKey();
      final LevelDBTypeInfo ti = getTypeInfo(valueClass);

      // Batching updates per type
      synchronized (ti) {
        final LevelDBTypeInfo.Index naturalIndex = ti.naturalIndex();
        final Collection<LevelDBTypeInfo.Index> indices = ti.indices();

        try (WriteBatch batch = db().createWriteBatch()) {
          while (valueIter.hasNext()) {
            final Object value = valueIter.next();

            assert serializedValueIter.hasNext();
            final byte[] serializedObject = serializedValueIter.next();

            Object existing;
            try {
              existing = get(naturalIndex.entityKey(null, value), valueClass);
            } catch (NoSuchElementException e) {
              existing = null;
            }

            PrefixCache cache = new PrefixCache(value);
            byte[] naturalKey = naturalIndex.toKey(naturalIndex.getValue(value));
            for (LevelDBTypeInfo.Index idx : indices) {
              byte[] prefix = cache.getPrefix(idx);
              idx.add(batch, value, existing, serializedObject, naturalKey, prefix);
            }
          }
          assert !serializedValueIter.hasNext();
          db().write(batch);
        }
      }
    }
  }
```

I think using multiple threads to write data to leveldb might shorten the switching time but it would introduce more overhead to SHS.

Reply:
It is a function of how loaded your disk is, iops it can sustain, txn's leveldb can do concurrently.

Reply:
Make sense. I am testing it on the mac which has SSD and the disk is not busy. I think the improvement might be more obvious on HDD or busy disk.  @HeartSaVioR @tgravescs Do we need to add batch write support for leveldb on this pr?

Reply:
I don't think it's mandatory. You can file another issue as "improvement" for this, but IMHO working with this is completely optional for you. I think we have already asked so many things to do.

Reply:
Sounds good. We can improve that afterward.